### ISSUE TITLE
Remove indent

### ISSUE THREAD

**Issue by ranebrown** [L3MON4D3/LuaSnip#436]
Is there a way to remove indent from where the snippet is triggered from? If I'm on a new line but indented on that line and type the trigger text and expand I would like it to insert the snippet starting on column 0 of that line instead of where the original text was typed.

```lua
    trig|

--> actual behavior
    this is the snippet
      contains some indent

--> desired behavior
this is the snippet
  contains some indent
```

**Comment by L3MON4D3** [comment]
You could use `regTrig` to match everything in the line before the trigger
```lua
s({trig=".*trig", regTrig=true}, {t{"this is the snippet","\tcontains some indent"}})
```
This should work, but there is some bug that still gives the snippet indent. I'll look into it.

**Comment by L3MON4D3** [comment]
Seems to work with 07cf1a1

**Comment by ranebrown** [comment]
In the example you posted should the `.*` be selected as a placeholder? I type `trig` expand and get `.*|trig` -> tab -> `.*trig|` and need to expand again to get the text with the correct indentation.

**Comment by L3MON4D3** [comment]
Oh that sounds weird, here's how it works for me:

https://user-images.githubusercontent.com/41961280/169680593-9e503dd0-b0fb-4e8a-a7c7-6e4829b8abff.mp4

**Comment by ranebrown** [comment]
Looks like that behavior is actually caused by cmp. If I exit the cmp menu before expanding the trigger it works as expected. Seems related to the discussion here https://github.com/saadparwaiz1/cmp_luasnip/issues/6. It seems like this should be fixable by using the [name field](https://github.com/L3MON4D3/LuaSnip/blob/07cf1a1fa01c7ecb2091d0cd04af2a2e70676058/doc/luasnip.txt#L143) in the snippet definition. Is the intention of that field to show that name instead of the `trig` text?

**Comment by ranebrown** [comment]
Tested a bit and forcing cmp_luasnip to use the name field doesn't solve the problem since it expands using [this helper](https://github.com/saadparwaiz1/cmp_luasnip/blob/a9de941bcbda508d0a45d28ae366bb3f08db2e36/lua/cmp_luasnip/init.lua#L131-L134) either way.

Here is a minimal init.lua if it helps but I'm guessing this will need to be fixed on the cmp_luasnip side.

<details>
  <summary>init.lua</summary>

```lua
vim.cmd([[set runtimepath=$VIMRUNTIME]])
vim.cmd([[set packpath=/tmp/nvim/site]])

local package_root = "/tmp/nvim/site/pack"
local install_path = package_root .. "/packer/start/packer.nvim"

local function load_plugins()
    require("packer").startup({
        function(use)
            use({ "wbthomason/packer.nvim" })
            use({ "L3MON4D3/LuaSnip" })
            use({ "saadparwaiz1/cmp_luasnip" })
            use({
                "hrsh7th/nvim-cmp",
                config = function()
                    local cmp = require("cmp")
                    local luasnip = require("luasnip")

                    cmp.setup({
                        snippet = {
                            expand = function(args)
                                require("luasnip").lsp_expand(args.body)
                            end,
                        },
                        mapping = {
                            ["<C-y>"] = cmp.mapping({
                                i = function(fallback)
                                    if cmp.visible() then
                                        cmp.confirm({
                                            behavior = cmp.ConfirmBehavior.Replace,
                                            select = false,
                                        })
                                    elseif luasnip.expandable() then
                                        luasnip.expand()
                                    else
                                        fallback()
                                    end
                                end,
                            }),
                            ["<C-e>"] = cmp.mapping({
                                i = cmp.mapping.abort(),
                            }),
                            ["<C-n>"] = cmp.mapping({
                                i = function()
                                    if luasnip.choice_active() then
                                        luasnip.change_choice(1)
                                    elseif cmp.visible() then
                                        cmp.select_next_item({
                                            behavior = cmp.SelectBehavior.Insert,
                                        })
                                    else
                                        cmp.complete()
                                    end
                                end,
                            }),
                            ["<C-p>"] = cmp.mapping({
                                i = function()
                                    if luasnip.choice_active() then
                                        luasnip.change_choice(-1)
                                    elseif cmp.visible() then
                                        cmp.select_prev_item({
                                            behavior = cmp.SelectBehavior.Insert,
                                        })
                                    else
                                        cmp.complete()
                                    end
                                end,
                            }),
                            ["<Tab>"] = cmp.mapping(function(fallback)
                                if luasnip.jumpable(1) then
                                    luasnip.jump(1)
                                else
                                    fallback()
                                end
                            end, { "i", "s" }),
                            ["<S-Tab>"] = cmp.mapping(function(fallback)
                                if luasnip.jumpable(-1) then
                                    luasnip.jump(-1)
                                else
                                    fallback()
                                end
                            end, { "i", "s" }),
                        },
                        sources = {
                            { name = "luasnip" },
                        },
                    })
                end,
            })
        end,
        config = {
            package_root = package_root,
            compile_path = install_path .. "/plugin/packer_compiled.lua",
        },
    })
end

if vim.fn.isdirectory(install_path) == 0 then
    vim.fn.system({
        "git",
        "clone",
        "https://github.com/wbthomason/packer.nvim",
        install_path,
    })
    load_plugins()
    require("packer").sync()
else
    load_plugins()
    require("packer").sync()
end
```
</details>

**Comment by L3MON4D3** [comment]
> In the example you posted should the `.*` be selected as a placeholder? I type `trig` expand and get `.*|trig` -> tab -> `.*trig|` and need to expand again to get the text with the correct indentation.

Ahh of course, the `expand_helper` expands a snippet for the pattern (this has to be done because the snippet might rely on some capture, which would not be populated when expanding the snippet via cmp), so the groups can be filled-in correctly. 

> I'm guessing this will need to be fixed on the cmp_luasnip side.

Im not sure if there really is anything to fix, cmp_luasnip provides a list of snippets, and one of those is expanded at the cursor. Matching the line against the trigger and removing text does not sound like the right thing to do.

There's another way to get that snippet: use the [indentSnippetNode](https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#indentsnippetnode), it can reset indentation of nodes inside it. This will, however not affect the line the snippet was expanded on, so that will need to be cleared, for example using the [snippet-callbacks](https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#snippets)

**Comment by L3MON4D3** [comment]
Okay, example didn't really work, the problem is that parts of the snippet are removed, which is never a good idea.

Instead, I added the `pre_expand`-callback (on brach `pre_expand_callback`) (example in #438) that should work even when expanded with nvim-cmp.

**Comment by ranebrown** [comment]
Tested that branch and it seems to be working for the snippet I have. Thanks!

**Comment by L3MON4D3** [comment]
Nice, I think I'll mehr it tomorrow or so

**Comment by L3MON4D3** [comment]
Merged! (#438)
